### DISCUSSION TITLE
Deprecate result.partition in favor of new list.partition_map

### DISCUSSION THREAD

**Discussion by SaphiraKai** [gleam-lang/gleam#3265] [Ideas & suggestions]
I would like to propose deprecating `result.partition` from the standard library, and superseding it with a more generic `list.partition_map` function.

My reasoning for suggesting this is:
- `result.partition` is only useful for `List(Result(a, e))`. It does not accept any other type, nor does it accept a callback to easily extend its usefulness beyond the simple case of collecting successes and failures into separate lists. `list.partition_map` (if implemented as suggested) is generic over `List(a)`, and does accept a callback to extend its behaviour without a separate call to `list.map`.
- `result.partition` being located in the `result` module rather than the `list` module makes it harder to find for those who might find use in it outside of the case mentioned above.
- `list.partition_map` would provide identical behaviour to `result.partition` when passed an identity function (e.g. `fn(a) { a }`) as callback.
- `result.partition` returns the lists in reverse order compared to the input list, requiring a separate call to `list.reverse` to correct the order. Regarding this, `list.partition_map` would provide consistent behaviour to the other `list` module functions.

For ease of reference, this is the current implementation of `result.partition`:
```gleam
pub fn partition(results: List(Result(a, e))) -> #(List(a), List(e)) {
  do_partition(results, [], [])
}

fn do_partition(results: List(Result(a, e)), oks: List(a), errors: List(e)) {
  case results {
    [] -> #(oks, errors)
    [Ok(a), ..rest] -> do_partition(rest, [a, ..oks], errors)
    [Error(e), ..rest] -> do_partition(rest, oks, [e, ..errors])
  }
}
```

And this is the implementation of `list.partition_map` that I propose:
```gleam
pub fn partition_map(
  list: List(a),
  with categorise: fn(a) -> Result(b, c),
) -> #(List(b), List(c)) {
  do_partition_map(list, categorise, [], [])
}

fn do_partition_map(list, categorise, oks, errors) {
  case list {
    [] -> #(reverse(oks) , reverse(errors)) 
    [x, ..xs] ->
      case categorise(x) {
        Ok(b) -> do_partition_map(xs, categorise, [b, ..oks], errors)
        Error(c) -> do_partition_map(xs, categorise, oks, [c, ..errors])
      }
  }
}
```

I've written this to be similar to `list.partition` and `list.filter_map`.

As this is my first issue on this repository, please let me know if there are any improvements I can make to this comment.

**Comment by lpil** [comment]
What's the motivation for deprecating this when we already have map and partition?

**Comment by SaphiraKai** [comment]
The reason I suggest deprecating `result.partition` is because `partition_map` would render it largely redundant, being able to do everything it can and then some.

<details>
<summary>Tangent about the performance of different approaches</summary>

A valid concern that was pointed out to me in the Discord is that `result.partition` not "correcting" the order of its output with an internal `list.reverse` call means it doesn't make a potentially performance-heavy decision on the developer's behalf, that the other `list` functions do.

Looking into this, I ran a few naive tests using `birl` to benchmark 3 different approaches to solving the same problem on various list sizes, and my overall finding was that using `result.partition` is actually almost always slower than the other approaches, regardless of the size of the list.
I also tested each method again, intentionally reversing the order, which adds an *additional* reverse to the other two pipelines and removes a reverse from the `result.partition` pipeline, and even then `result.partition` was only faster on lists on the order of tens of millions of elements and still barely outperformed the other two. In fact, at smaller list sizes it was somehow *slower* than the other two despite having *no* calls to reverse, compared to 2 calls negating eachother in the other pipelines.

But that's probably a less relevant tangent anyway. If you actually care about those results I'd be happy to clean up and publish the code I'm testing with.
</details>

Personally the reason I care about `partition_map` is because this:
```gleam
[#("wibble", 0), #("wobble", 1), #("wabble", 2)]
  |> partition_map(fn(a) {
    case a.1 |> int.is_odd {
      True -> Ok(a.0)
      False -> Error(a.1)
    }
  })
```
Is simpler than:
```gleam
[#("wibble", 0), #("wobble", 1), #("wabble", 2)]
    |> list.map(fn(a) {
      case a.1 |> int.is_odd {
        True -> Ok(a.0)
        False -> Error(a.1)
      }
    })
    |> list.reverse
    |> result.partition
```
And significantly simpler than:
```gleam
[#("wibble", 0), #("wobble", 1), #("wabble", 2)]
    |> list.partition(fn(a) { a.1 |> int.is_odd })
    |> pair.map_first(fn(evens) { evens |> list.map(pair.first) })
    |> pair.map_second(fn(odds) { odds |> list.map(pair.second) })
```
The 2nd approach being the slowest and counterintuitively requiring a `list.reverse`, and the 3rd requiring you import `pair` and define 3 anonymous functions to perform what I would argue is conceptually a single operation.

The 3rd approach also becomes significantly more cumbersome if you're partitioning a list of `Result`s rather than a tuple, because instead of mapping with `pair.first` and `pair.second` you'd have to map with two `let assert`s to destructure the `Result`:
```gleam
[Ok(0), Error(1), Ok(2)]
  |> list.partition(result.is_ok)
  |> pair.map_first(fn(oks) {
    oks
    |> list.map(fn(a) {
      let assert Ok(inner) = a
      inner
    })
  })
  |> pair.map_second(fn(errors) {
    errors
    |> list.map(fn(e) {
      let assert Error(inner) = e
      inner
    })
  })
```
Of course, for this you'd probably just use `list.reverse` followed by`result.partition` instead. Or you could use `partition_map` and simply pass an identity function:
```gleam
[Error(0), Ok("wobble"), Error(2)] |> partition_map(fn(a) { a })
```

# TL;DR
It's my understanding that Gleam strives to provide a minimal number of tools, and that each of those tools be highly adaptable and widely applicable. I believe `partition_map` to be a good candidate for such.

I apologize for the text wall, I'm just trying to be as clear and thorough as possible on my reasoning.

**Comment by SaphiraKai** [comment]
If you don't believe there's sufficient reason to deprecate `result.partition`, you could also treat this as a proposal for an additional function rather than a replacement.

**Comment by lpil** [comment]
I don't personally find the "simpler" examples given here more compelling than the other examples, but perhaps it's still useful. Perhaps it should be a tuple of a bool and a value rather than a result as it's unrelated to failure.

**Comment by SaphiraKai** [comment]
I was following the precedent set by `filter_map`, although I do agree. Is there a reason we couldn't/shouldn't have an `Either` type for cases like this? A tuple as you suggested would work fine as well.

**Comment by SaphiraKai** [comment]
On second thought, a pair would limit the output lists to being the same type. You'd need a `#(Bool, a, b)` to resolve that.

**Comment by lpil** [comment]
Good point on the return type!

I'm going to convert this to a discussion as there's no clear way to move forward yet, and the demand is also not yet clear.
